spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/BigQueryFileMessageHandler.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2017-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.bigquery.integration;
+
+import java.io.BufferedInputStream;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.channels.Channels;
+
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.FormatOptions;
+import com.google.cloud.bigquery.JobInfo.WriteDisposition;
+import com.google.cloud.bigquery.TableDataWriteChannel;
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.WriteChannelConfiguration;
+
+import org.springframework.integration.handler.AbstractReplyProducingMessageHandler;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHandlingException;
+import org.springframework.util.StreamUtils;
+
+/**
+ * Outbound channel adapter which handles sending and loading files to a BigQuery table.
+ */
+public class BigQueryFileMessageHandler extends AbstractReplyProducingMessageHandler {
+
+	private final BigQuery bigQuery;
+
+	private boolean autoDetectSchema = true;
+
+	private WriteDisposition writeDisposition = WriteDisposition.WRITE_APPEND;
+
+	public BigQueryFileMessageHandler(BigQuery bigQuery) {
+		this.bigQuery = bigQuery;
+	}
+
+	@Override
+	protected Object handleRequestMessage(Message<?> message) {
+		try {
+			String datasetName = message.getHeaders().get(BigQuerySpringMessageHeaders.DATASET_NAME, String.class);
```

Comment:
Interesting choice to use the headers for this rather than properties of the actual channel adapter.
Maybe we can use both, where headers take precedence if set.
I'm curious @artembilan's opinion on this.

Reply:
See `PubSubMessageHandler` for its `topicExpression` as a sample how to be.

It is indeed must not be a strict requirement for just header population: we can obtain required info from any other places. That's where a SpEL with its evaluation context comes to the help. It helps us to avoid header-enricher upfront.

Reply:
Done. I've changed it so the user can set a default for the message handler options; however the message can provide headers which will override the default settings.